### PR TITLE
Move2Plan/Execute (MoveTo Behavior Overhaul)

### PR DESCRIPTION
**Pull request by egordon**
# Description

Adds the following functionality / overhauled behaviors:

* MoveIt2Plan: given a bunch of planning parameters returns a JointTrajectory
* * If all goal constraints are satisfied, return None and Success
* * If any path constraint is unsatisfied, return None and Failure (unless explicitly ignored)
* MoveIt2XConstraint: compile blackboard objects into the constraint kwargs dictionary expected by MoveIt2Plan
* MoveIt2Execute: Given a JointTrajectory, execute it on the robot
* Add the option for setting the debug log_level at the launch file level
* Add the default action_select policy to the ada_feeding main launch file
* Extend the AcquireFood tree to carry out MoveAbove in order to test the basic functionality of the new behaviors.

# Testing procedure

Follow the testing procedure in #102 with the following additional steps:
* Verify that the robot moves 5cm above the shown `food` frame with the tines perpendicular to the X (red) axis (i.e. constant action 0)
* Alternate between the 2 carrot images and see the robot go back and forth moving above the two foods

_Note: will be happy to take recommendations on how to test other components of the MoveIt2 behaviors, but this is sufficient for me now to continue working on the AcquireFood tree_

# Before opening a pull request
- [x] Format your code using [black formatter](https://black.readthedocs.io/en/stable/) `python3 -m black .`
- [x] Run your code through [pylint](https://pylint.readthedocs.io/en/latest/) and address all warnings/errors. The only warnings that are acceptable to not address is TODOs that should be addressed in a future PR. From the top-level `ada_feeding` directory, run: `pylint --recursive=y --rcfile=.pylintrc .`.

# Before Merging
- [ ] `Squash & Merge`

### CODE REVIEW

File: ada_feeding/ada_feeding/behaviors/acquisition/compute_action_constraints.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/acquisition/compute_action_constraints.py:87]: Nit: add a logger message here as well

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py:77]: FWIW, you can use `typing.Annotated` to include metadata in a type hint. The metadata won't be checked by a static type checker, but will be additional information available to the reader of the code. That might be a bit easier to understand than a comment (here and wherever else you use a comment in `blackboard_outputs`). https://docs.python.org/3/library/typing.html#typing.Annotated

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py:108]: Nit: log error (here and in all below constraints)

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py:273]: In the docstring, link to this for more details on parameterization: https://github.com/ros-planning/moveit_msgs/blob/master/msg/OrientationConstraint.msg

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: I don't think this is percent of the geodesic, I think it is literally distance in meters. https://github.com/ros-planning/moveit_msgs/blob/9756797dcbc293c7cae66541a7bc0c7acec85892/srv/GetCartesianPath.srv#L22 https://github.com/ros-planning/moveit2/blob/5dd0d5e63c9dae08976656cb1c8473ca7a34c058/moveit_core/robot_state/src/cartesian_interpolator.cpp#L116

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: This note is particularly important, because `allowed_planning_time` is for the planner itself, but MoveGroup has additional steps it takes before and after the planner, plus there is communication time with the BT. If programmers do use a Timeout decorator, `allowed_planning_time` should be **_less than the Timeout value_**, else it will likely necessarily fail. Perhaps add a brief mention of that here?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]:
```suggestion
        max_velocity_scale: [0,1] determines max joint velocity of the trajectory
```

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]:
```suggestion
        cartesian_jump_threshold: If >0, large relative joint jumps between cartesian planner IKs
```

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:111]: No need to do this, but given how many parameters are cartesian-specific, we should consider creating a CartesianPlanConfig class for it. If you agree, maybe we can add a TODO for it in this docstring?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:218]: Nit: I typically prefer only blocking for the line(s) where the global resource is being used, to free it up for other threads to use.

For example, the `tf_lock` does not need to be occupied when getting the trajectory and neither the moveit2 not tf locks need to be occupied while scaling the velocity for cartesian trajectories. Freeing up these locks would allow e.g., a behavior that is listening to the TF tree in parallel to not be blocked for longer than necessary.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:296]: Perhaps log a warning here with a throttle, to make it easier for programmers to debug while their code is looping?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:301]: Not sure where to add it, but in some docstring it would be helpful to specify that all the goal constraints we set specify a single goal region (e.g., we take the `and` of all the goal constraints). This library doesn't currently allow multiple goal regions (e.g., the `or` of multiple constraints), even though [it is allowed in MoveIt2](https://github.com/ros-planning/moveit_msgs/blob/9756797dcbc293c7cae66541a7bc0c7acec85892/msg/MotionPlanRequest.msg#L15).

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Won't this happen in `terminate` anyway, which is called as soon as this block returns success.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Consider getting this a single time before the for loop, to save multiple blackboard lookup calls if it is not a local/constant.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:296]: Same with the path constraints

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: This double negative ("not ignore") took quite a while to parse, although I have now verified that it is correct. 

Perhaps a more readable/interpretable variant of this code would be:

```
elif ignore_violated_path_constraints:
    continue # act as if this constraint did not exist
else:
    paths_satisfied = False
    break
```

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Given that `end_joint_state` is `Optional`, should we set it to None here?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: I'm confused why you are only checking the `index` in `curr_joint_state` in certain cases -- there is no guarentee that the joint_names in the constraint will be ordered in the same way as the joint names in the state publisher.

I'd recommend simplifying this logic by:

1. Looping over the min len of `joint_positions` and `joint_name` (remove unnecessary conditional)
2. For every joint_name, find its corresponding index in joint_state and use that to get `curr_position`.
3. (++) Log an error and return False if `index` throws an exception.
4. (Nit) I don't think `des_position` is the most clear name, maybe `dest_position`?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Unnecessary in this function?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: My last commit to `pymoveit2` addressed this: https://github.com/personalrobotics/pymoveit2/blob/amaln/move_collision/pymoveit2/moveit2.py#L1717

Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Can be removed from orientatio constraints as well

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Since we did `fabs` in orientation constraints, perhaps we should do it in position and joint constraints as well to provide uniform functionality?

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:602]: Perhaps do the same for `import csv`? I don't think we use that anywhere but here

Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:602]: And for `from ament_index_python.packages import get_package_share_directory`

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py:74]: Unclear to me what the numbers here refer to. Because it is a union (not an int), anyone using this will first have to check what type it is an instance of, and then once then know that they just refer to the message definition for the error codes, right? The below comments don't clarify anything for me.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py]: Do we really want to handle both these cases the same? The trajectory not existing on the blackboard seems like a programmer error that should be flagged, whereas a trajectory of None is where constraints were already satisfied, so it should succee.

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/trees/acquire_food_tree.py:109]: Even if we are switching from the decorator to behavior convention for defining constraints, we should still have a `ClearConstraints` behavior. Else, if this tree is run multiple times (e.g., to acquire multiple bites), it will keep appending the same pose constraint to the list of goal constraints.

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (amalnanavati) [inline ada_feeding/ada_feeding/trees/acquire_food_tree.py:62]: We should coalesce on a style for large composites. I personally prefer defining each behavior separately and then either adding them to the sequence at the end ([as in `MoveToMouth`](https://github.com/personalrobotics/ada_feeding/blob/9f5c219391930651baa852727004ba3c6fecba20/ada_feeding/ada_feeding/trees/move_to_mouth_tree.py#L438)), or appending them to the children as you go ([as in `scoped_behavior`](https://github.com/personalrobotics/ada_feeding/blob/9f5c219391930651baa852727004ba3c6fecba20/ada_feeding/ada_feeding/idioms/scoped_behavior.py#L88), and [other idioms in `py_trees`](https://github.com/splintered-reality/py_trees/blob/devel/py_trees/idioms.py#L195)). My concern with the style you are using is too many nested lists and too much indentation.

**Review by amalnanavati — APPROVED** [review]
LGTM, some changes should be made but nothing that requires another review.

Re. your question on testing the other aspects of `MoveIt2Plan/Execute`, once this is merged in I'll percolate the changes to the other behaviors and trees, which should test the other aspects of it.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py:74]: The error codes are both type int, I'll update.
It's a case where MoveItErrorCodes and GoalStatus are msg objects with constant ints defined inside. 

Basically, the return can be either, which works because when there's an error GoalStatus reports a positive number and Move it reports a negative number.

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (egordon) [reply ada_feeding/ada_feeding/trees/acquire_food_tree.py:109]: (1) Perhaps we should consider having a helper function at a higher level that clears the blackboard before each tree run

(2) I can add this for convenience, but note that the behavior takes both an input and an output constraints array, so if the input constraints are unset or None, then it will create a new one to write to the blackboard, solving the multiple runs problem

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (egordon) [reply ada_feeding/ada_feeding/trees/acquire_food_tree.py:62]: (my larger problem with this whole issue is that they're isn't a clean way to define a tree, XML does this better)

The problem with append_children is that it obfuscates the tree structure for anybody reading the code.

What I could do is define the leaf behaviors /subtrees separately, but putting them all in at the end makes it easier to browse (especially if you use an IDE that allows collapsing if code blocks, which is most of them).

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:218]: The problem is that with statement need to be preceded by an "if locked return running" to prevent blocking the tick thread.

But that could cause update to return prematurely and lose or recalculate local state.

I think it's better to treat the entire update, which should be quick anyway, with a single with block to prevent this.

Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py:218]: Good point, I agree with the way you’ve currently written the code.

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (amalnanavati) [reply ada_feeding/ada_feeding/trees/acquire_food_tree.py:109]: (1) I don’t think we need a higher level function, because most blackboard inputs are reset in initialise. `constraints` is the notable exception, because an initialise call can’t differentiate between the case where it is getting called on a new tree, or the constraint was populated by prior behaviors.
(2) I’m not sure how this solves the multiple runs problem? I agree that if constraints is under it will initialize it. The multiple runs issue is that constraints will already be set on a second run, so it will keep appending to that.
(3) another way to address this could be to give each constraint a name/key, and make the constraint container stored on the blackboards a dict. That way it will override the constraint if it’s name was already there.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py
Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py:74]: I don’t think they are both ints. To get to the Int, users will have to call `error_code.status` if it is a GoalStatus, or `error_code.Val` if it is a MoveItErrorCode. Hence, users are already conditioning on the type of error code.

Comment (amalnanavati) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py:74]: EDIT: I think we are saying the same thing. I just felt like your comment made it more confusing, but I understand your perspective that pointing out in the commend that one case is always positive and the other is negative can be helpful. I think just clarifying the comment should resolve this.

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_constraints.py:77]: Right now, I don't think there's a ton of value-add

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: Yeah, looks like meters, but since it is only a single value, for rotations, steps there are 2deg/1cm.

https://github.com/ros-planning/moveit2/blob/498a9f3d659cbcf0db03cef584d4ca16468e4dab/moveit_core/robot_state/include/moveit/robot_state/cartesian_interpolator.h#L79

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: des_position == desired_position

Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: I'm confused. I'm not making the assumption that constraint "joint_names" is the same as the publisher. that's what the call to `index` is for, and it happens every time the constraint joint_names is defined at the current index.

The problem with the proposed logic is that I don't think it handles the logic as-defined. If "joint_names"[i] exists, do the re-mapping. Otherwise, just borrow the order from the state publisher.

(In practice, for clarity, joint_names should be defined every time, but this code handles the case where it isn't).

Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py]: I can simplify the logic by requiring joint_names to be either len==0 (or None) or len(joint_positions)

File: ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py
Comment (egordon) [reply ada_feeding/ada_feeding/behaviors/moveit2/moveit2_execute.py:74]: Just made sure that error_code is always the full type and not an int (I made that mistake once by returning `MoveItErrorCodes.SUCCESS`)

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (egordon) [reply ada_feeding/ada_feeding/trees/acquire_food_tree.py:109]: Discussed externally, by passing in None as a constant to the first Move2PoseConstraints, we solve the multi-run problem

File: ada_feeding/ada_feeding/trees/acquire_food_tree.py
Comment (egordon) [reply ada_feeding/ada_feeding/trees/acquire_food_tree.py:62]: No update on this right now, but the full AcquireFoodTree will likely define leaf behaviors separately.

**Pull request merged**
